File: trunk/java-api/src/de/zib/scalaris/Scalaris.java
/*
 *  Copyright 2007-2011 Zuse Institute Berlin
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
package de.zib.scalaris;

import java.io.IOException;

import com.ericsson.otp.erlang.OtpAuthException;
import com.ericsson.otp.erlang.OtpErlangAtom;
import com.ericsson.otp.erlang.OtpErlangExit;
import com.ericsson.otp.erlang.OtpErlangInt;
import com.ericsson.otp.erlang.OtpErlangList;
import com.ericsson.otp.erlang.OtpErlangLong;
import com.ericsson.otp.erlang.OtpErlangObject;
import com.ericsson.otp.erlang.OtpErlangString;
import com.ericsson.otp.erlang.OtpErlangTuple;

/**
 * Provides methods to read and write key/value pairs to a scalaris ring.
 * 
 * <p>
 * Each operation is a single transaction. If you are looking for more
 * transactions, use the {@link Transaction} class instead.
 * </p>
 * 
 * <p>
 * Instances of this class can be generated using a given connection to a
 * scalaris node using {@link #Scalaris(Connection)} or without a
 * connection ({@link #Scalaris()}) in which case a new connection is
 * created using {@link ConnectionFactory#createConnection()}.
 * </p>
 * 
 * <p>
 * There are two paradigms for reading and writing values:
 * <ul>
 *  <li> using Java {@link String}s: {@link #read(String)}, {@link #write(String, String)}
 *       <p>This is the safe way of accessing scalaris where type conversions
 *       are handled by the API and the user doesn't have to worry about anything else.</p>
 *       <p>Be aware though that this is not the most efficient way of handling strings!</p>
 *  <li> using custom {@link OtpErlangObject}s: {@link #readObject(OtpErlangString)},
 *       {@link #writeObject(OtpErlangString, OtpErlangObject)}
 *       <p>Here the user can specify custom behaviour and increase performance.
 *       Handling the stored types correctly is at the user's hand.</p>
 *       <p>An example using erlang objects to improve performance for inserting strings is
 *       provided by {@link de.zib.scalaris.examples.CustomOtpFastStringObject} and can be
 *       tested by {@link de.zib.scalaris.examples.FastStringBenchmark}.</p>
 * </ul> 
 * </p>
 * 
 * <h3>Reading values</h3>
 * <pre>
 * <code style="white-space:pre;">
 *   String key;
 *   OtpErlangString otpKey;
 *   
 *   Scalaris sc = new Scalaris();
 *   String value             = sc.read(key);          // {@link #read(String)}
 *   OtpErlangObject optValue = sc.readObject(otpKey); // {@link #readObject(OtpErlangString)}
 * </code>
 * </pre>
 * 
 * <p>For the full example, see {@link de.zib.scalaris.examples.ScalarisReadExample}</p>
 * 
 * <h3>Writing values</h3>
 * <pre>
 * <code style="white-space:pre;">
 *   String key;
 *   String value;
 *   OtpErlangString otpKey;
 *   OtpErlangString otpValue;
 *   
 *   Scalaris sc = new Scalaris();
 *   sc.write(key, value);             // {@link #write(String, String)}
 *   sc.writeObject(otpKey, otpValue); // {@link #writeObject(OtpErlangString, OtpErlangObject)}
 * </code>
 * </pre>
 * 
 * <p>For the full example, see {@link de.zib.scalaris.examples.ScalarisWriteExample}</p>
 * 
 * <h3>Deleting values</h3>
 * <pre>
 * <code style="white-space:pre;">
 *   String key;
 *   int timeout;
 *   DeleteResult result;
 *   
 *   Scalaris sc = new Scalaris();
 *   sc.delete(key);                    // {@link #delete(String)}
 *   sc.delete(key, timeout);           // {@link #delete(String, int)}
 *   result = sc.getLastDeleteResult(); // {@link #getLastDeleteResult()}
 * </code>
 * </pre>
 * 
 * <h3>Connection errors</h3>
 * 
 * Errors when setting up connections or trying to send/receive RPCs will be
 * handed to the {@link ConnectionPolicy} that has been set when the connection
 * was created. By default, {@link ConnectionFactory} uses
 * {@link DefaultConnectionPolicy} which implements automatic connection-retries
 * by classifying nodes as good or bad depending on their previous state. The
 * number of automatic retries is adjustable (default: 3).
 * 
 * @author Nico Kruber, kruber@zib.de
 * @version 2.5
 * @since 2.0
 */
public class Scalaris {
	/**
	 * Connection to a Scalaris node.
	 */
	private Connection connection;
	
	/**
	 * Stores the result list returned by erlang during a delete operation.
	 * 
	 * @see #delete(String)
	 */
	private OtpErlangList lastDeleteResult = null;
	
	/**
	 * Constructor, uses the default connection returned by
	 * {@link ConnectionFactory#createConnection()}.
	 * 
	 * @throws ConnectionException
	 *             if the connection fails
	 */
	public Scalaris() throws ConnectionException {
		connection = ConnectionFactory.getInstance().createConnection();
	}

	/**
	 * Constructor, uses the given connection to an erlang node.
	 * 
	 * @param conn
	 *            connection to use for the transaction
	 * 
	 * @throws ConnectionException
	 *             if the connection fails
	 */
	public Scalaris(Connection conn) throws ConnectionException {
		connection = conn;
	}
	
	// /////////////////////////////
	// read methods
	// /////////////////////////////
	
	/**
	 * Gets the value stored under the given <tt>key</tt>.
	 * 
	 * @param key
	 *            the key to look up
	 * 
	 * @return the value stored under the given <tt>key</tt>
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to fetch the value
	 * @throws NotFoundException
	 *             if the requested key does not exist
	 * @throws UnknownException
	 *             if any other error occurs
	 */
	public OtpErlangObject readObject(OtpErlangString key)
            throws ConnectionException, TimeoutException, NotFoundException,
            UnknownException {
		OtpErlangObject received_raw = null;
		try {
			received_raw = connection.doRPC("api_tx", "read",
					new OtpErlangList(key));
			OtpErlangTuple received = (OtpErlangTuple) received_raw;

			/*
			 * possible return values:
			 *  {ok, Value} | {fail, timeout | not_found}
			 */
			if (received.elementAt(0).equals(CommonErlangObjects.failAtom)) {
				OtpErlangObject reason = received.elementAt(1);
				if (reason.equals(CommonErlangObjects.timeoutAtom)) {
					throw new TimeoutException(received_raw);
				} else if (reason.equals(CommonErlangObjects.notFoundAtom)) {
					throw new NotFoundException(received_raw);
				} else {
					throw new UnknownException(received_raw);
				}
			} else if (received.elementAt(0).equals(CommonErlangObjects.okAtom) && received.arity() == 2) {
                OtpErlangObject value = received.elementAt(1);
                return value;
			} else {
                throw new UnknownException(received_raw);
			}
		} catch (OtpErlangExit e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (OtpAuthException e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (IOException e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (ClassCastException e) {
			// e.printStackTrace();
			// received_raw is not null since the first class cast is after the RPC!
			throw new UnknownException(e, received_raw);
		}
	}

	/**
	 * Gets the value stored under the given <tt>key</tt>.
	 * 
	 * @param key
	 *            the key to look up
	 * 
	 * @return the (string) value stored under the given <tt>key</tt>
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to fetch the value
	 * @throws NotFoundException
	 *             if the requested key does not exist
	 * @throws UnknownException
	 *             if any other error occurs
	 * 
	 * @see #readObject(OtpErlangString)
	 */
	public String read(String key) throws ConnectionException,
			TimeoutException, NotFoundException, UnknownException {
		try {
			CustomOtpStringObject result = new CustomOtpStringObject();
			readCustom(key, result);
			return result.getValue();
//			return ((OtpErlangString) readObject(new OtpErlangString(key)))
//					.stringValue();
		} catch (ClassCastException e) {
			// e.printStackTrace();
			throw new UnknownException(e);
		}
	}
	
	/**
	 * Gets the value stored under the given <tt>key</tt>.
	 * 
	 * @param key
	 *            the key to look up
	 * @param value 
	 *            container that stores the value returned by scalaris
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to fetch the value
	 * @throws NotFoundException
	 *             if the requested key does not exist
	 * @throws UnknownException
	 *             if any other error occurs
	 * 
	 * @see #readObject(OtpErlangString)
	 * @since 2.1
	 */
	public void readCustom(String key, CustomOtpObject<?> value)
            throws ConnectionException, TimeoutException, NotFoundException,
            UnknownException {
		try {
			value.setOtpValue(readObject(new OtpErlangString(key)));
		} catch (ClassCastException e) {
			// e.printStackTrace();
			throw new UnknownException(e);
		}
	}

	// /////////////////////////////
	// write methods
	// /////////////////////////////

	/**
	 * Stores the given <tt>key</tt>/<tt>value</tt> pair.
	 * 
	 * @param key
	 *            the key to store the value for
	 * @param value
	 *            the value to store
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to write the value
     * @throws AbortException
     *             if the commit of the write failed
	 * @throws UnknownException
	 *             if any other error occurs
	 */
	public void writeObject(OtpErlangString key, OtpErlangObject value)
			throws ConnectionException, TimeoutException, AbortException, UnknownException {
		OtpErlangObject received_raw = null;
		try {
			received_raw = connection.doRPC("api_tx", "write",
					new OtpErlangList(new OtpErlangObject[] { key, value }));
			OtpErlangTuple received = (OtpErlangTuple) received_raw;

            /*
             * possible return values:
             *  {ok} | {fail, timeout | abort}
             */
            if (received.elementAt(0).equals(CommonErlangObjects.failAtom)) {
                OtpErlangObject reason = received.elementAt(1);
                if (reason.equals(CommonErlangObjects.timeoutAtom)) {
                    throw new TimeoutException(received_raw);
                } else if (reason.equals(CommonErlangObjects.abortAtom)) {
                    throw new AbortException(received_raw);
                } else {
                    throw new UnknownException(received_raw);
                }
            } else if (received.equals(CommonErlangObjects.okTupleAtom)) {
                return;
            } else {
                throw new UnknownException(received_raw);
            }
		} catch (OtpErlangExit e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (OtpAuthException e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (IOException e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (ClassCastException e) {
			// e.printStackTrace();
			// received_raw is not null since the first class cast is after the RPC!
			throw new UnknownException(e, received_raw);
		}
	}

	/**
	 * Stores the given <tt>key</tt>/<tt>value</tt> pair.
	 * 
	 * @param key
	 *            the key to store the value for
	 * @param value
	 *            the value to store
	 *
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to write the value
     * @throws AbortException
     *             if the commit of the write failed
	 * @throws UnknownException
	 *             if any other error occurs
	 *
	 * @see #writeObject(OtpErlangString, OtpErlangObject)
	 */
	public void write(String key, String value) throws ConnectionException,
			TimeoutException, AbortException, UnknownException {
		writeCustom(key, new CustomOtpStringObject(value));
//		writeObject(new OtpErlangString(key), new OtpErlangString(value));
	}
	
	/**
	 * Stores the given <tt>key</tt>/<tt>value</tt> pair.
	 * 
	 * @param key
	 *            the key to store the value for
	 * @param value
	 *            the value to store
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to write the value
     * @throws AbortException
     *             if the commit of the write failed
	 * @throws UnknownException
	 *             if any other error occurs
	 * 
	 * @see #writeObject(OtpErlangString, OtpErlangObject)
	 * @since 2.1
	 */
	public void writeCustom(String key, CustomOtpObject<?> value)
			throws ConnectionException, TimeoutException, AbortException, UnknownException {
		writeObject(new OtpErlangString(key), value.getOtpValue());
	}
	
	// /////////////////////////////
	// delete methods
	// /////////////////////////////
	
	/**
	 * Tries to delete all replicas of the given <tt>key</tt> in 2000ms.
	 * 
	 * @param key
	 *            the key to delete
	 * 
	 * @return the number of successfully deleted replicas
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to delete the value
	 * @throws NodeNotFoundException
	 *             if no scalaris node was found
	 * @throws UnknownException
	 *             if any other error occurs
	 * 
	 * @since 2.2
	 * 
	 * @see #delete(String, int)
	 */
	public long delete(String key) throws ConnectionException,
	TimeoutException, UnknownException, NodeNotFoundException {
		return delete(key, 2000);
	}

	/**
	 * Tries to delete all replicas of the given <tt>key</tt>.
	 * 
	 * WARNING: This function can lead to inconsistent data (e.g. deleted items
	 * can re-appear). Also when re-creating an item the version before the
	 * delete can re-appear.
	 * 
	 * @param key
	 *            the key to delete
	 * @param timeout
	 *            the time (in milliseconds) to wait for results
	 * 
	 * @return the number of successfully deleted replicas
	 * 
	 * @throws ConnectionException
	 *             if the connection is not active or a communication error
	 *             occurs or an exit signal was received or the remote node
	 *             sends a message containing an invalid cookie
	 * @throws TimeoutException
	 *             if a timeout occurred while trying to delete the value
	 * @throws NodeNotFoundException
	 *             if no scalaris node was found
	 * @throws UnknownException
	 *             if any other error occurs
	 * 
	 * @since 2.2
	 * 
	 * @see #delete(String)
	 */
	public long delete(String key, int timeout) throws ConnectionException,
	TimeoutException, UnknownException, NodeNotFoundException {
		OtpErlangObject received_raw = null;
		lastDeleteResult = null;
		try {
			received_raw = connection.doRPC("transaction_api", "delete",
					new OtpErlangList( new OtpErlangObject[] {
							new OtpErlangString(key),
							new OtpErlangInt(timeout) }));
			OtpErlangTuple received = (OtpErlangTuple) received_raw;

			/*
			 * possible return values:
			 *  - {ok, pos_integer(), list()}
			 *  - {fail, timeout}
			 *  - {fail, timeout, pos_integer(), list()}
			 *  - {fail, node_not_found}
			 */
			if (received.elementAt(0).equals(new OtpErlangAtom("fail"))) {
				OtpErlangObject reason = received.elementAt(1);
				if (reason.equals(new OtpErlangAtom("timeout"))) {
					if (received.arity() > 2) {
						lastDeleteResult = (OtpErlangList) received.elementAt(3);
					}
					throw new TimeoutException(received_raw);
				} else if (reason.equals(new OtpErlangAtom("node_not_found"))) {
					throw new NodeNotFoundException(received_raw);
				} else {
					throw new UnknownException(received_raw);
				}
			} else {
				lastDeleteResult = (OtpErlangList) received.elementAt(2);
				long succeeded = ((OtpErlangLong) received.elementAt(1)).longValue();
				return succeeded;
			}
		} catch (OtpErlangExit e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (OtpAuthException e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (IOException e) {
			// e.printStackTrace();
			throw new ConnectionException(e);
		} catch (ClassCastException e) {
			// e.printStackTrace();
			// received_raw is not null since the first class cast is after the RPC!
			throw new UnknownException(e, received_raw);
		}
	}

	/**
	 * Returns the result of the last call to {@link #delete(String)}.
	 * 
	 * NOTE: This function traverses the result list returned by erlang and
	 * therefore takes some time to process. It is advised to store the returned
	 * result object once generated.
	 * 
	 * @return the delete result
	 * 
	 * @throws UnknownException
	 *             is thrown if an unknown reason was encountered
	 * 
	 * @see #delete(String)
	 */
	public DeleteResult getLastDeleteResult() throws UnknownException {
		try {
			return new DeleteResult(lastDeleteResult);
		} catch (UnknownException e) {
			throw new UnknownException(e, lastDeleteResult);
		}
	}
	
	/**
	 * Closes the transaction's connection to a scalaris node.
	 * 
	 * Note: Subsequent calls to the other methods will throw
	 * {@link ConnectionException}s!
	 */
	public void closeConnection() {
		connection.close();
	}
}

File: python-api/Scalaris.py
#!/usr/bin/python
# Copyright 2011 Zuse Institute Berlin
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.

import json, httplib, urlparse,  socket
import sys
import base64

default_url = 'http://localhost:8000/jsonrpc.yaws'
default_timeout = 5 # socket timeout in seconds

def _json_call(conn,  uri,  function, params):
    params = {'version': '1.1',
              'method': function,
              'params': params,
              'id': 0}
    #print params
    # use compact JSON encoding:
    params_json = json.dumps(params, separators=(',',':'))
    #print params_json
    # no need to quote - we already encode to json:
    headers = {"Content-type": "application/json"}
    try:
        #conn.request("POST", uri.path, urllib.quote(params_json), headers)
        conn.request("POST", uri.path, params_json, headers)
        response = conn.getresponse()
        #print response.status, response.reason
        if (response.status < 200 or response.status >= 300):
            raise ConnectionException(response)
        data = response.read()
    except socket.timeout,  (instance):
        raise ConnectionException(instance)
    #print data
    response_json = json.loads(data)
    #print response_json
    return response_json['result']

def _json_encode_value(value):
    if isinstance(value,  bytearray):
        return {'type': 'as_bin',  'value': base64.b64encode(value)}
    else:
        return {'type': 'as_is',  'value': value}

def _json_decode_value(value):
    if ('type' not in value) or ('value' not in value):
        raise UnknownException(value)
    if value['type'] == 'as_bin':
        return bytearray(base64.b64decode(value['value']),  'base64')
    else:
        return value['value']

def getConnection(url,  timeout = default_timeout):
    uri = urlparse.urlparse(url)
    return httplib.HTTPConnection(uri.hostname,  uri.port, timeout = timeout)
    
# result: {'status': 'ok'} or {'status': 'ok', 'value': xxx} or
#         {'status': 'fail', 'reason': 'timeout' or 'abort' or 'not_found}
# op: 'read' or 'write' or 'commit', 'test_and_set' or 'any'
def _process_result(result,  op):
    if isinstance(result,  dict) and 'status' in result:
        if result['status'] == 'ok':
            if 'value' in result:
                return _json_decode_value(result['value'])
            elif op == 'read':
                raise UnknownException((result,  op))
        elif result['status'] == 'fail' and 'reason' in result:
            if result['reason'] == 'timeout':
                raise TimeoutException(result)
            elif result['reason'] == 'not_found' and op not in ['commit',  'write']:
                raise NotFoundException(result)
            elif result['reason'] == 'abort' and op not in ['read',  'write']:
                raise AbortException(result)
            elif result['reason'] == 'key_changed' and op == 'test_and_set' and 'value' in result:
                raise KeyChangedException(result,  _json_decode_value(result['value']))
            else:
                raise UnknownException(result)
        else:
            raise UnknownException(result)
    else:
        raise UnknownException(result)

# Exception that is thrown if a the commit of a write operation on a
# scalaris ring fails.
class AbortException(Exception):
   def __init__(self, raw_result):
       self.raw_result = raw_result
   def __str__(self):
       return repr(self.raw_result)

# Exception that is thrown if an operation on a scalaris ring fails because
# a connection does not exist or has been disconnected.
class ConnectionException(Exception):
   def __init__(self, raw_result):
       self.raw_result = raw_result
   def __str__(self):
       return repr(self.raw_result)

# Exception that is thrown if a test_and_set operation on a scalaris ring
# fails because the old value did not match the expected value.
class KeyChangedException(Exception):
   def __init__(self, raw_result,  old_value):
       self.raw_result = raw_result
       self.old_value = old_value
   def __str__(self):
       return repr(self.raw_result) + ', old value: ' + repr(self.old_value)

# Exception that is thrown if a delete operation on a scalaris ring fails
# because no scalaris node was found.
class NodeNotFoundException(Exception):
   def __init__(self, raw_result):
       self.raw_result = raw_result
   def __str__(self):
       return repr(self.raw_result)

# Exception that is thrown if a read operation on a scalaris ring fails
# because the key did not exist before.
class NotFoundException(Exception):
   def __init__(self, raw_result):
       self.raw_result = raw_result
   def __str__(self):
       return repr(self.raw_result)

# Exception that is thrown if a read or write operation on a scalaris ring
#fails due to a timeout.
class TimeoutException(Exception):
   def __init__(self, raw_result):
       self.raw_result = raw_result
   def __str__(self):
       return repr(self.raw_result)

# Generic exception that is thrown during operations on a scalaris ring, e.g.
# if an unknown result has been returned.
class UnknownException(Exception):
   def __init__(self, raw_result):
       self.raw_result = raw_result
   def __str__(self):
       return repr(self.raw_result)

class TransactionSingleOp(object):
    def __init__(self, url = None,  conn = None,  timeout = default_timeout):
        if (url == None):
            self._uri = urlparse.urlparse(default_url)
        else:
            self._uri = urlparse.urlparse(url)
        
        if (conn == None):
            try:
                self._conn = httplib.HTTPConnection(self._uri.hostname,  self._uri.port,  timeout = timeout)
            except Exception,  (instance):
                raise ConnectionException(instance)
        else:
            self._conn = conn

    def read(self,  key):
        result = _json_call(self._conn,  self._uri, 'read', [key])
        # results: {'status': 'ok', 'value', xxx} or
        #          {'status': 'fail', 'reason': 'timeout' or 'not_found'}
        return _process_result(result,  'read')

    def write(self,  key, value):
        value = _json_encode_value(value)
        result = _json_call(self._conn,  self._uri, 'write', [key, value])
        # results: {'status': 'ok'} or
        #          {'status': 'fail', 'reason': 'timeout' or 'abort'}
        _process_result(result,  'commit')
    
    def testAndSet(self,  key, oldvalue, newvalue):
        oldvalue = _json_encode_value(oldvalue)
        newvalue = _json_encode_value(newvalue)
        result = _json_call(self._conn,  self._uri, 'test_and_set', [key, oldvalue, newvalue])
        # results: {'status': 'ok'} or
        #          {'status': 'fail', 'reason': 'timeout' or 'abort' or 'not_found'} or
        #          {'status': 'fail', 'reason': 'key_changed', 'value': xxx}
        _process_result(result,  'test_and_set')

    def nop(self,  value):
        value = _json_encode_value(value)
        _json_call(self._conn,  self._uri, 'nop', [value])
        # results: 'ok'
    
    def closeConnection(self):
        self._conn.close()

class Transaction(object):
    def __init__(self, url = None,  conn = None,  timeout = default_timeout):
        if (url == None):
            self._uri = urlparse.urlparse(default_url )
        else:
            self._uri = urlparse.urlparse(url)
        
        if (conn == None):
            try:
                self._conn = httplib.HTTPConnection(self._uri.hostname,  self._uri.port,  timeout = timeout)
            except Exception,  (instance):
                raise ConnectionException(instance)
        else:
            self._conn = conn
        
        self._tlog = None
    
    def req_list(self,  reqlist):
        if self._tlog == None:
            result = _json_call(self._conn,  self._uri, 'req_list', [reqlist.getJSONRequests()])
        else:
            result = _json_call(self._conn,  self._uri, 'req_list', [self._tlog, reqlist.getJSONRequests()])
        # results: {'tlog': xxx,
        #           'results': [{'status': 'ok'} or {'status': 'ok', 'value': xxx} or
        #                       {'status': 'fail', 'reason': 'timeout' or 'abort' or 'not_found}]}
        if 'tlog' not in result or 'results' not in result or not isinstance(result['results'],  list) or len(result['results']) < 1:
            raise UnknownException(result)
        self._tlog = result['tlog']
        return result['results']
    
    # result: {'status': 'ok'} or {'status': 'ok', 'value': xxx} or
    #         {'status': 'fail', 'reason': 'timeout' or 'abort' or 'not_found}
    # op: 'read' or 'write' or 'commit' or 'any'
    def process_result(self,  result,  op):
        return _process_result(result,  op)
    
    # results: {'status': 'ok'} or
    #          {'status': 'fail', 'reason': 'timeout' or 'abort'}
    def commit(self):
        result = self.req_list(ReqList().addCommit())[0]
        self.process_result(result,  'commit')
        # reset tlog (minor optimization which is not done in req_list):
        self._tlog = None
    
    # results: None
    def abort(self):
        self._tlog = None
    
    # results: {'status': 'ok', 'value', xxx} or
    #              {'status': 'fail', 'reason': 'timeout' or 'not_found'}
    def read(self,  key):
        result = self.req_list(ReqList().addRead(key))[0]
        return self.process_result(result,  'read')
    
    # results: {'status': 'ok'} or
    #              {'status': 'fail', 'reason': 'timeout' or 'abort'}
    def write(self,  key,  value):
        result = self.req_list(ReqList().addWrite(key,  value))[0]
        self.process_result(result,  'commit')

    def nop(self,  value):
        value = _json_encode_value(value)
        _json_call(self._conn,  self._uri, 'nop', [value])
        # results: 'ok'
    
    def closeConnection(self):
        self._conn.close()

class ReqList(object):
    def __init__(self):
        self.requests = []
    
    def addRead(self,  key):
        self.requests.append({'read': key})
        return self
    
    def addWrite(self,  key,  value):
        self.requests.append({'write': {key: _json_encode_value(value)}})
        return self
    
    def addCommit(self):
        self.requests.append({'commit': 'commit'})
        return self
    
    def getJSONRequests(self):
        return self.requests

class PubSub(object):
    def __init__(self, url = None,  conn = None,  timeout = default_timeout):
        if (url == None):
            self._uri = urlparse.urlparse(default_url)
        else:
            self._uri = urlparse.urlparse(url)
        
        if (conn == None):
            try:
                self._conn = httplib.HTTPConnection(self._uri.hostname,  self._uri.port,  timeout = timeout)
            except Exception,  (instance):
                raise ConnectionException(instance)
        else:
            self._conn = conn

    def publish(self,  topic,  content):
        content = _json_encode_value(content)
        result = _json_call(self._conn,  self._uri, 'publish', [topic,  content])
        # results: {'status': 'ok'} or
        #          {'status': 'fail', 'reason': 'timeout' or 'abort'}
        _process_result(result,  'commit')

    def subscribe(self,  topic,  url):
        url = _json_encode_value(url)
        result = _json_call(self._conn,  self._uri, 'subscribe', [topic,  url])
        # results: {'status': 'ok'} or
        #          {'status': 'fail', 'reason': 'timeout' or 'abort'}
        _process_result(result,  'commit')

    def unsubscribe(self,  topic,  url):
        url = _json_encode_value(url)
        result = _json_call(self._conn,  self._uri, 'unsubscribe', [topic,  url])
        # results: {'status': 'ok'} or
        #          {'status': 'fail', 'reason': 'timeout' or 'abort'}
        _process_result(result,  'commit')

    def getSubscribers(self,  topic):
        result = _json_call(self._conn,  self._uri, 'get_subscribers', [topic])
        # results: {'status': 'ok', 'value', xxx} or
        #          {'status': 'fail', 'reason': 'timeout' or 'not_found'}
        return _process_result(result,  'read')

    def nop(self,  value):
        value = _json_encode_value(value)
        _json_call(self._conn,  self._uri, 'nop', [value])
        # results: 'ok'
    
    def closeConnection(self):
        self._conn.close()
    
# if the expected value is a list, the returned value could by (mistakenly) a string if it is a list of integers
# -> provide a method for converting such a string to a list
def str_to_list(value):
    if (isinstance(value, str) or isinstance(value, unicode)):
        chars = list(value)
        return [ord(char) for char in chars]
    else:
        return value

if __name__ == "__main__":
    import sys
    sc = TransactionSingleOp()
    if (sys.argv[1] == "read"):
        key = sys.argv[2]
        try:
            value = sc.read(key)
            print 'read(' + key + ') = ' + rep(value)
        except ConnectionException,  (instance):
            print 'read(' + key + ') failed with connection error'
            sys.exit(1)
        except TimeoutException,  (instance):
            print 'read(' + key + ') failed with timeout'
            sys.exit(1)
        except NotFoundException,  (instance):
            print 'read(' + key + ') failed with not_found'
            sys.exit(1)
        except UnknownException,  (instance):
            print 'read(' + key + ') failed with unknown: ' + str(instance)
            sys.exit(1)
    elif (sys.argv[1] == "write"):
        key = sys.argv[2]
        value = sys.argv[3]
        try:
            sc.write(key,  value)
            print 'write(' + key + ', ' + value + '): ok'
        except ConnectionException,  (instance):
            print 'write(' + key + ', ' + value + ') failed with connection error'
            sys.exit(1)
        except TimeoutException,  (instance):
            print 'write(' + key + ', ' + value + ') failed with timeout'
            sys.exit(1)
        except AbortException,  (instance):
            print 'write(' + key + ', ' + value + ') failed with abort'
            sys.exit(1)
        except UnknownException,  (instance):
            print 'write(' + key + ', ' + value + ') failed with unknown: ' + str(instance)
            sys.exit(1)
